src/datepicker/datepicker.js

```diff
@@ -127,6 +127,16 @@ angular.module('ui.bootstrap.datepicker', ['ui.bootstrap.dateparser', 'ui.bootst
     return arrays;
   };
 
+  // Fix a hard-reprodusible bug with timezones
+  // The bug depends on OS, browser, current timezone and current date
+  // i.e.
+  // var date = new Date(2014, 0, 1);
+  // console.log(date.getFullYear(), date.getMonth(), date.getDate(), date.getHours());
+  // can result in "2013 11 31 23" because of the bug.
+  this.fixTimeZone = function(date) {
+    date.setHours(date.getHours() > 12 ? date.getHours() + 2 : 0);
```

Comment:
Today reproduced it on Win 7 with GMT-0300 (Brasilia) using Firefox 36.0.4.
Code:
```javascript
var date = new Date(2016, 9, 16);

console.log('date: ', date);
console.log('hours:', date.getHours());
console.log('timestamp:', date.getTime());

date.setHours(date.getHours() + 1);

console.log('date: ', date);
console.log('hours:', date.getHours());
console.log('timestamp:', date.getTime());

date.setHours(date.getHours() + 2);

console.log('date: ', date);
console.log('hours:', date.getHours());
console.log('timestamp:', date.getTime());
```
Result:
```
date: Date {Sat Oct 15 2016 23:00:00 GMT-0300}
hours: 23
timestamp: 1476583200000

date: Date {Sat Oct 15 2016 23:00:00 GMT-0300}
hours: 23
timestamp: 1476583200000

date: Date {Sun Oct 16 2016 01:00:00 GMT-0200}
hours: 1
timestamp: 1476586800000
```
So we should make sure that hours are not equal to 23. If they are, we need to add two hours to the date (as you can see adding one hour doesn't work in this case).
If this part misleads you: `date.getHours() > 12 ` I can change to `date.getHours() === 23 ` .

I think behavior of this fix is quite predictable and should deal with all the cases of this issue.

Sorry if I'm not clear enough.